Blog post 'Testing Kafka Connectors' (#45)
gunnarmorling/morling.dev: content/blog/testing-kafka-connectors.asciidoc

---
title: "Testing Kafka Connectors"
date: 2022-08-25T09:20:00+01:00
draft: false
markup: adoc
---
:source-highlighter: rouge
:rouge-style: base16.dark
:icons: font
:imagesdir: /images
ifdef::env-github[]
:imagesdir: ../../static/images
endif::[]

https://kafka.apache.org/documentation/#connect[Kafka Connect] is a key factor for the wide-spread adoption of Apache Kafka:
a framework and runtime environment for connectors,
it makes the task of getting data either into Kafka or out of Kafka solely a matter of configuration,
rather than a bespoke programming job.
There's dozens, if not hundreds, of readymade source and sink connectors,
allowing you to create no-code data pipelines between all kinds of databases, APIs, and other systems.

There may be situations though where there is no existing connector matching your requirements,

> in which case you can implement your own custom connector using the Kafka Connect framework.

Reviewer: Maybe link here to https://kafka.apache.org/documentation/#connect_development ?

Author: Good idea, will do.

[source,java,linenums=true]
----
public class EtcdSourceConnectorTest {

  @Test
  public void shouldCreateConfigurationForTasks() throws Exception {
    EtcdSourceConnector connector = new EtcdSourceConnector();
    Map<String, String> config = new HashMap<>();
    config.put(
      "clusters", 
      "etcd-a=http://etcd-a-1:2379,http://etcd-a-2:2379,http://etcd-a-3:2379;etcd-b=http://etcd-b-1:2379;etcd-c=http://etcd-c-1:2379"
    ); // <1>
    connector.start(config);

    List<Map<String, String>> taskConfigs = connector.taskConfigs(2); // <2>
    assertThat(taskConfigs).hasSize(2);

    // <3>
    taskConfig = taskConfigs.get(0);
    assertThat(taskConfig).containsEntry("clusters",
        "etcd-a=http://etcd-a-1:2379,http://etcd-a-2:2379,http://etcd-a-3:2379;etcd-b=http://etcd-b-1:2379");

    // <4>
    taskConfig = taskConfigs.get(1);
    assertThat(taskConfig).containsEntry("clusters",
        "etcd-c=http://etcd-c-1:2379");
  }
}
----
<1> Configure the connector with three Etcd clusters
<2> Request the configuration for two tasks
<3> The first task should handle the first two clusters
<4> The second task should handle the remaining third cluster

> Things looks similar for testing the actual polling loop of the connector's task class.

Reviewer: ```suggestion
Things look similar when testing the actual polling loop of the connector's task class.
```

As this is about testing a source connector,
we first need to do some data changes in the configured Etcd cluster(s),

> before we can assert the corresponding ``SourceRecord``s emitted by the task.

Reviewer: ```suggestion
before we can assert the corresponding ``SourceRecord``s that are emitted by the task.
```

As part of kc-etcd, I've implemented a _very_ basic testing harness named https://github.com/gunnarmorling/kcetcd/tree/main/src/test/java/dev/morling/kcute[kcute]

> ("Kafka Connect Unit Testing") which simplifies that task a bit.

Reviewer: Maybe replace the word "task" with "process" since the word task comes up a lot already when talking about connect.

Author: LOL, yeah, good point. Changing it.

<1> Set up an Etcd cluster using the JUnit extension provided by the https://github.com/etcd-io/jetcd[jetcd] client project
<2> Set up the task unter test using kcute
<3> Obtain a client for Etcd and do some changes
<4> Retrieve three records for the specified topic via kcute
<5> Assert the emitted ``SourceRecord``s corresponding to the data changes done before in Etcd

This testing approach works very well in general;
in particular it doesn't require you to start Apache Kafka (and ZooKeeper) nor Kafka Connect,
resulting in very fast test execution times and a great dev experience when creating and running these tests in your IDE.

But there are some limitations, too.
Essentially, we end up emulating the behavior of the actual Kafka Connect runtime in our testing harness.
This can become tedious when more advanced Connect features are required for a given test,
for instance retrying/restart logic, the dynamic reconfiguration of connector tasks while the connector is running, etc.
Ideally, there'd be a testing harness with all these capabilities provided as part of Kafka Connect itself
(similar in spirit to the `TopologyTestDriver` of Kafka Streams),
but in the absence of that,

> we may be better of for certain tests by deploying our source connector into an actual Kafka Connect instance and run assertions against the topic(s) it writes to.

Reviewer: ```suggestion
we may be better off, for certain tests, deploying our source connector into an actual Kafka Connect instance and running assertions against the topic(s) it writes to.
```

== Integration Tests

When it comes to setting up the required infrastructure for integration tests in Java,
the go-to solution these days is the excellent https://www.testcontainers.org/[Testcontainers] project.

> So let's see what it takes for testing a custom Kafka connector using Testcontainers.

Reviewer: ```suggestion
So let's see what it takes to test a custom Kafka connector using Testcontainers.
```

As far as Kafka itself is concerned, there's a https://www.testcontainers.org/modules/kafka/[module] for that coming with Testcontainers,
based on Confluent Platform.
Alternatively, you could use the https://github.com/strimzi/test-container[Testcontainers module] from the Strimzi project,

> providing you with plain upstream Apache Kafka container images.

Reviewer: ```suggestion
which provides you with plain upstream Apache Kafka container images.
```

For Kafka Connect, we provide a https://debezium.io/documentation/reference/stable/integrations/testcontainers.html[Testcontainers integration] as part of the Debezium project,
offering an API for registering connectors and controlling their lifecycle.

Now, unfortunately, the application server like deployment model of Kafka Connect poses a challenge when it comes to testing a connector which is built as part of the current project itself.
For each connector plug-in, Connect expects a directory on its plug-in path which contains all the JARs of the connector itself and its dependencies.
I'm not aware of any kind of "exploded mode",
where you could point Connect to a directory which contains a connector's class files and its dependencies in JAR form.

I.e. packaging the connector into a JAR must happen as part of the test preparation.
In order to make integration tests friendly towards being run from within an IDE,
this should happen programmatically within the test itself.

> The entire code for doing this is a bit too long (and boring) for sharing it in this blog post,

Reviewer: ```suggestion
The entire code for doing this is a bit too long (and boring) to share in this blog post,
```

    connectContainer.registerConnector("my-connector", connector);
    connectContainer.ensureConnectorTaskState("my-connector", 0, State.RUNNING);

    KV kvClient = client.getKVClient();

    // <7>
    // insert
    ByteSequence key = ByteSequence.from("key-1".getBytes());
    ByteSequence value = ByteSequence.from("value-1".getBytes());
    kvClient.put(key, value).get();

    // update
    key = ByteSequence.from("key-1".getBytes());
    value = ByteSequence.from("value-1a".getBytes());
    kvClient.put(key, value).get();

    // delete
    key = ByteSequence.from("key-2".getBytes());
    kvClient.delete(key).get();

    // <8>
    List<ConsumerRecord<String, String>> records = drain(getConsumer(kafkaContainer), 3);

    // insert
    ConsumerRecord<String, String> record = records.get(0);
    assertThat(record.key()).isEqualTo("key-1");
    assertThat(record.value()).isEqualTo("value-1");

    // update
    record = records.get(1);
    assertThat(record.key()).isEqualTo("key-1");
    assertThat(record.value()).isEqualTo("value-1a");

    // delete
    record = records.get(2);
    assertThat(record.key()).isEqualTo("key-2");
    assertThat(record.value()).isNull();
  }
}
----

> <1> Set up Apache Kafka

Reviewer: ```suggestion
<1> Set up Apache Kafka using the Testcontainers Kafka module
```

<2> Set up Kafka Connect, mounting the _target/kcetcd-connector_ directory onto the plug-in path; as part of the project's Maven build, all the dependencies of the kcetcd connector are copied into that directory

> <3> Set up Etcd

Reviewer: Is this also using a Testcontainer module?

Author: Yes, it is; will clarify.

verifying its behavior when run via Kafka Connect.

== Wrap-Up

In this post, we've discussed two approaches for testing Kafka Connect source connectors:
plain unit tests, "manually" invoking the methods of the connector/task classes under test,
and integration tests, deploying a connector into Kafka Connect and verifying its behavior there via Testcontainers.

The former approach provides you with faster turnaround times and shorter feedback cycles,
whereas the latter approach gives you the confidence of testing a connector within the actual Kafka Connect runtime environment,
at the cost of a more complex infrastructure set-up and longer test execution times.
While we've focused on testing source connectors in this post,
both approaches could equally be applied to sink connectors,
with the only difference being that you'd feed records to the connector (either directly or by writing to a Kafka topic) and then observe and assert the expected state changes of the sink system in question.

You can find the complete source code for this post, including some parts omitted here for brevity,
in the https://github.com/gunnarmorling/kcetcd[kc-etcd] repository on GitHub.

> If you think that having a test harness like kcute for unit testing connectors is a good idea and it's something you'd like to contribute to, then please let me know.

Reviewer: ```suggestion
If you think that having a test harness like [kcute](https://github.com/gunnarmorling/kcetcd/tree/main/src/test/java/dev/morling/kcute[kcute]) for unit testing connectors is a good idea and it's something you'd like to contribute to, then please let me know.
```
